halo2_proofs/src/poly/commitment/msm.rs

```diff
@@ -20,23 +22,32 @@ impl<'a, C: CurveAffine> MSM<'a, C> {
         let g_scalars = None;
         let w_scalar = None;
         let u_scalar = None;
-        let other_scalars = vec![];
-        let other_bases = vec![];
+        let other = BTreeMap::new();
 
         MSM {
             params,
             g_scalars,
             w_scalar,
             u_scalar,
-            other_scalars,
-            other_bases,
+            other,
         }
     }
 
     /// Add another multiexp into this one
     pub fn add_msm(&mut self, other: &Self) {
-        self.other_scalars.extend(other.other_scalars.iter());
-        self.other_bases.extend(other.other_bases.iter());
+        for (x, (scalar, y)) in other.other.iter() {
+            self.other
+                .entry(*x)
+                .and_modify(|(our_scalar, our_y)| {
+                    if our_y == y {
+                        *our_scalar += *scalar;
+                    } else {
+                        assert!(*our_y == -*y);
+                        *our_scalar -= *scalar;
+                    }
+                })
+                .or_insert((*scalar, *y));
```

Comment:
Non-blocking: this could be de-duplicated from `add_msm` and `append_term` with a helper method.

Reply:
Yes, I couldn't quite get that to work, so deferred it for a later cleanup.